File: backend/app/api/compliance.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, HttpUrl
from typing import List, Optional, Dict, Any
import logging

from app.core.compliance import (
    get_compliance_manager, 
    ComplianceCheck, 
    ComplianceLevel,
    check_url_compliance
)
from app.api.deps import get_current_active_user
from app.schemas.user import User

logger = logging.getLogger(__name__)

router = APIRouter()


class ComplianceCheckRequest(BaseModel):
    """コンプライアンスチェックリクエスト"""
    url: HttpUrl
    compliance_level: Optional[str] = "moderate"


class ComplianceCheckResponse(BaseModel):
    """コンプライアンスチェックレスポンス"""
    url: str
    allowed: bool
    warnings: List[str]
    errors: List[str]
    recommendations: List[str]
    delay_seconds: float
    compliance_level: str


class SitePolicyResponse(BaseModel):
    """サイトポリシーレスポンス"""
    url: str
    robots_txt_url: str
    terms_of_service_url: Optional[str]
    allows_crawling: bool
    requires_delay: float
    detected_restrictions: List[str]


class ComplianceStatsResponse(BaseModel):
    """コンプライアンス統計レスポンス"""
    total_checks: int
    allowed_count: int
    blocked_count: int
    warning_count: int
    domains_with_restrictions: List[str]
    average_delay: float


@router.post("/check", response_model=ComplianceCheckResponse)
async def check_compliance(
    request: ComplianceCheckRequest,
    current_user: User = Depends(get_current_active_user)
):
    """
    指定されたURLのコンプライアンスをチェック
    """
    try:
        compliance_manager = get_compliance_manager()
        
        # コンプライアンスレベルを設定
        if request.compliance_level:
            try:
                level = ComplianceLevel(request.compliance_level.lower())
                compliance_manager.compliance_level = level
            except ValueError:
                raise HTTPException(
                    status_code=400,
                    detail=f"無効なコンプライアンスレベル: {request.compliance_level}"
                )
        
        # コンプライアンスチェック実行
        check_result = await compliance_manager.check_compliance(str(request.url))
        
        return ComplianceCheckResponse(
            url=str(request.url),
            allowed=check_result.allowed,
            warnings=check_result.warnings,
            errors=check_result.errors,
            recommendations=check_result.recommendations,
            delay_seconds=check_result.delay_seconds,
            compliance_level=compliance_manager.compliance_level.value
        )
        
    except Exception as e:
        logger.error(f"コンプライアンスチェックエラー: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"コンプライアンスチェックに失敗しました: {str(e)}"
        )


@router.get("/site-policy/{domain:path}", response_model=SitePolicyResponse)
async def get_site_policy(
    domain: str,
    current_user: User = Depends(get_current_active_user)
):
    """
    指定されたドメインのサイトポリシーを取得
    """
    try:
        # URLの正規化
        if not domain.startswith(('http://', 'https://')):
            domain = f"https://{domain}"
        
        compliance_manager = get_compliance_manager()
        policy = compliance_manager.get_site_policy(domain)
        
        # 制限の検出
        detected_restrictions = []
        if not policy.allows_crawling:
            detected_restrictions.append("robots.txt制限")
        if policy.requires_delay > 2:
            detected_restrictions.append(f"高遅延要求({policy.requires_delay}秒)")
        if policy.terms_of_service_url:
            detected_restrictions.append("利用規約検出")
        
        return SitePolicyResponse(
            url=domain,
            robots_txt_url=policy.robots_txt_url,
            terms_of_service_url=policy.terms_of_service_url,
            allows_crawling=policy.allows_crawling,
            requires_delay=policy.requires_delay,
            detected_restrictions=detected_restrictions
        )
        
    except Exception as e:
        logger.error(f"サイトポリシー取得エラー: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"サイトポリシーの取得に失敗しました: {str(e)}"
        )


@router.post("/batch-check")
async def batch_check_compliance(
    urls: List[HttpUrl],
    compliance_level: str = "moderate",
    current_user: User = Depends(get_current_active_user)
):
    """
    複数URLのコンプライアンスを一括チェック
    """
    if len(urls) > 50:  # 制限
        raise HTTPException(
            status_code=400,
            detail="一度にチェックできるURLは50個までです"
        )
    
    try:
        compliance_manager = get_compliance_manager()
        
        # コンプライアンスレベル設定
        try:
            level = ComplianceLevel(compliance_level.lower())
            compliance_manager.compliance_level = level
        except ValueError:
            raise HTTPException(
                status_code=400,
                detail=f"無効なコンプライアンスレベル: {compliance_level}"
            )
        
        results = []
        for url in urls:
            try:
                check_result = await compliance_manager.check_compliance(str(url))
                results.append({
                    "url": str(url),
                    "allowed": check_result.allowed,
                    "warnings": check_result.warnings,
                    "errors": check_result.errors,
                    "recommendations": check_result.recommendations,
                    "delay_seconds": check_result.delay_seconds
                })
            except Exception as e:
                results.append({
                    "url": str(url),
                    "allowed": False,
                    "warnings": [],
                    "errors": [f"チェック失敗: {str(e)}"],
                    "recommendations": [],
                    "delay_seconds": 0
                })
        
        return {"results": results}
        
    except Exception as e:
        logger.error(f"バッチコンプライアンスチェックエラー: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"バッチチェックに失敗しました: {str(e)}"
        )


@router.get("/stats", response_model=ComplianceStatsResponse)
async def get_compliance_stats(
    current_user: User = Depends(get_current_active_user)
):
    """
    コンプライアンス統計を取得
    """
    try:
        compliance_manager = get_compliance_manager()
        
        # 統計情報を計算
        total_checks = len(compliance_manager.site_policies)
        domains_with_restrictions = []
        total_delay = 0
        
        for domain, policy in compliance_manager.site_policies.items():
            if not policy.allows_crawling or policy.requires_delay > 2 or policy.terms_of_service_url:
                domains_with_restrictions.append(domain)
            total_delay += policy.requires_delay
        
        average_delay = total_delay / max(total_checks, 1)
        
        return ComplianceStatsResponse(
            total_checks=total_checks,
            allowed_count=total_checks - len(domains_with_restrictions),
            blocked_count=len(domains_with_restrictions),
            warning_count=len(domains_with_restrictions),  # 簡易計算
            domains_with_restrictions=[d.replace('https://', '').replace('http://', '') for d in domains_with_restrictions],
            average_delay=round(average_delay, 2)
        )
        
    except Exception as e:
        logger.error(f"コンプライアンス統計取得エラー: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"統計の取得に失敗しました: {str(e)}"
        )


@router.post("/record-result")
async def record_request_result(
    url: HttpUrl,
    success: bool,
    current_user: User = Depends(get_current_active_user)
):
    """
    リクエスト結果を記録（バックオフ戦略用）
    """
    try:
        compliance_manager = get_compliance_manager()
        compliance_manager.record_request_result(str(url), success)
        
        return {
            "message": "リクエスト結果を記録しました",
            "url": str(url),
            "success": success
        }
        
    except Exception as e:
        logger.error(f"結果記録エラー: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"結果の記録に失敗しました: {str(e)}"
        )


@router.get("/recommended-headers/{domain:path}")
async def get_recommended_headers(
    domain: str,
    current_user: User = Depends(get_current_active_user)
):
    """
    指定ドメインの推奨HTTPヘッダーを取得
    """
    try:
        if not domain.startswith(('http://', 'https://')):
            domain = f"https://{domain}"
        
        compliance_manager = get_compliance_manager()
        headers = compliance_manager.get_recommended_headers(domain)
        
        return {
            "url": domain,
            "recommended_headers": headers
        }
        
    except Exception as e:
        logger.error(f"推奨ヘッダー取得エラー: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"推奨ヘッダーの取得に失敗しました: {str(e)}"
        ) 

File: frontend/src/utils/compliance.ts
/**
 * フロントエンドコンプライアンス機能
 * バックエンドのコンプライアンスAPIとの連携
 */

import { SecureHttpClient } from './security'

export interface ComplianceCheck {
  url: string
  allowed: boolean
  warnings: string[]
  errors: string[]
  recommendations: string[]
  delay_seconds: number
  compliance_level: string
}

export interface SitePolicy {
  url: string
  robots_txt_url: string
  terms_of_service_url: string | null
  allows_crawling: boolean
  requires_delay: number
  detected_restrictions: string[]
}

export interface ComplianceStats {
  total_checks: number
  allowed_count: number
  blocked_count: number
  warning_count: number
  domains_with_restrictions: string[]
  average_delay: number
}

export type ComplianceLevel = 'strict' | 'moderate' | 'permissive'

export class ComplianceManager {
  private baseUrl = '/api/compliance'

  /**
   * URLのコンプライアンスをチェック
   */
  async checkCompliance(
    url: string, 
    complianceLevel: ComplianceLevel = 'moderate'
  ): Promise<ComplianceCheck> {
    try {
      const response = await SecureHttpClient.fetch(`${this.baseUrl}/check`, {
        method: 'POST',
        body: JSON.stringify({
          url,
          compliance_level: complianceLevel
        })
      })

      if (!response.ok) {
        throw new Error(`コンプライアンスチェック失敗: ${response.status}`)
      }

      return await response.json()
    } catch (error) {
      console.error('Compliance check error:', error)
      throw error
    }
  }

  /**
   * サイトポリシーを取得
   */
  async getSitePolicy(domain: string): Promise<SitePolicy> {
    try {
      // ドメインをエンコード
      const encodedDomain = encodeURIComponent(domain)
      
      const response = await SecureHttpClient.fetch(
        `${this.baseUrl}/site-policy/${encodedDomain}`
      )

      if (!response.ok) {
        throw new Error(`サイトポリシー取得失敗: ${response.status}`)
      }

      return await response.json()
    } catch (error) {
      console.error('Site policy fetch error:', error)
      throw error
    }
  }

  /**
   * 複数URLの一括コンプライアンスチェック
   */
  async batchCheckCompliance(
    urls: string[], 
    _complianceLevel: ComplianceLevel = 'moderate'
  ): Promise<{ results: ComplianceCheck[] }> {
    try {
      const response = await SecureHttpClient.fetch(`${this.baseUrl}/batch-check`, {
        method: 'POST',
        body: JSON.stringify(urls),
        headers: {
          'Content-Type': 'application/json'
        }
      })

      if (!response.ok) {
        throw new Error(`バッチチェック失敗: ${response.status}`)
      }

      return await response.json()
    } catch (error) {
      console.error('Batch compliance check error:', error)
      throw error
    }
  }

  /**
   * コンプライアンス統計を取得
   */
  async getStats(): Promise<ComplianceStats> {
    try {
      const response = await SecureHttpClient.fetch(`${this.baseUrl}/stats`)

      if (!response.ok) {
        throw new Error(`統計取得失敗: ${response.status}`)
      }

      return await response.json()
    } catch (error) {
      console.error('Compliance stats error:', error)
      throw error
    }
  }

  /**
   * リクエスト結果を記録
   */
  async recordRequestResult(url: string, success: boolean): Promise<void> {
    try {
      await SecureHttpClient.fetch(`${this.baseUrl}/record-result`, {
        method: 'POST',
        body: JSON.stringify({ url, success })
      })
    } catch (error) {
      console.error('Record request result error:', error)
      // エラーを無視（統計目的のため）
    }
  }

  /**
   * 推奨HTTPヘッダーを取得
   */
  async getRecommendedHeaders(domain: string): Promise<Record<string, string>> {
    try {
      const encodedDomain = encodeURIComponent(domain)
      
      const response = await SecureHttpClient.fetch(
        `${this.baseUrl}/recommended-headers/${encodedDomain}`
      )

      if (!response.ok) {
        throw new Error(`推奨ヘッダー取得失敗: ${response.status}`)
      }

      const result = await response.json()
      return result.recommended_headers
    } catch (error) {
      console.error('Recommended headers error:', error)
      throw error
    }
  }
}

/**
 * コンプライアンス警告コンポーネント用のユーティリティ
 */
export class ComplianceWarningHelper {
  static getSeverityLevel(check: ComplianceCheck): 'error' | 'warning' | 'info' {
    if (!check.allowed || check.errors.length > 0) {
      return 'error'
    }
    if (check.warnings.length > 0) {
      return 'warning'
    }
    return 'info'
  }

  static getSeverityColor(severity: 'error' | 'warning' | 'info'): string {
    switch (severity) {
      case 'error':
        return 'text-red-600 bg-red-50 border-red-200'
      case 'warning':
        return 'text-yellow-600 bg-yellow-50 border-yellow-200'
      case 'info':
        return 'text-blue-600 bg-blue-50 border-blue-200'
    }
  }

  static getSeverityIcon(severity: 'error' | 'warning' | 'info'): string {
    switch (severity) {
      case 'error':
        return `<svg className="w-5 h-5" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 8v4m0 4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z" />
        </svg>`
      case 'warning':
        return `<svg className="w-5 h-5" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16c-.77.833.192 2.5 1.732 2.5z" />
        </svg>`
      case 'info':
        return `<svg className="w-5 h-5" fill="none" stroke="currentColor" viewBox="0 0 24 24">
          <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M13 16h-1v-4h-1m1-4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z" />
        </svg>`
    }
  }

  static formatDelay(seconds: number): string {
    if (seconds < 1) {
      return '即座'
    } else if (seconds < 60) {
      return `${seconds}秒`
    } else {
      const minutes = Math.floor(seconds / 60)
      const remainingSeconds = seconds % 60
      return remainingSeconds > 0 
        ? `${minutes}分${remainingSeconds}秒`
        : `${minutes}分`
    }
  }

  static getComplianceSummary(check: ComplianceCheck): string {
    if (!check.allowed) {
      return 'アクセス禁止'
    }
    
    const issueCount = check.warnings.length + check.errors.length
    if (issueCount === 0) {
      return 'コンプライアンス適合'
    }
    
    return `${issueCount}件の注意事項`
  }
}

/**
 * コンプライアンスチェック結果の表示フック
 */
export function useComplianceDisplay(check: ComplianceCheck | null) {
  if (!check) {
    return null
  }

  const severity = ComplianceWarningHelper.getSeverityLevel(check)
  const color = ComplianceWarningHelper.getSeverityColor(severity)
  const summary = ComplianceWarningHelper.getComplianceSummary(check)
  const formattedDelay = ComplianceWarningHelper.formatDelay(check.delay_seconds)

  return {
    severity,
    color,
    summary,
    formattedDelay,
    hasIssues: check.warnings.length > 0 || check.errors.length > 0,
    isBlocked: !check.allowed
  }
}

// シングルトンインスタンス
export const complianceManager = new ComplianceManager() 